klippy/extras/bed_mesh.py

```diff
@@ -0,0 +1,617 @@
+            % (param[0], config.get(*param)))
+    elif len(pair) == 1:
+        pair = (pair[0], pair[0])
+    if minval is not None:
+        if pair[0] < minval or pair[1] < minval:
+            raise config.error(
+                "Option '%s' in section bed_mesh must have a minimum of %s"
+                % (param[0]), minval)
+    if maxval is not None:
+        if pair[0] > maxval or pair[1] > maxval:
+            raise config.error(
+                "Option '%s' in section bed_mesh must have a maximum of %s"
+                % (param[0]), str(minval))
+    return pair
+
+
+class BedMesh:
+    FADE_DISABLE = 0x7FFFFFFF
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.z_adjust = 0.
+        self.calibrate = BedMeshCalibrate(config, self)
+        self.z_mesh = None
+        self.toolhead = None
+        self.horizontal_move_z = config.getfloat('horizontal_move_z', 5.)
+        fade_start = config.getfloat('fade_start', 1., minval=0.)
+        fade_end = config.getfloat('fade_end', 10.)
+        self.gcode = self.printer.lookup_object('gcode')
+        self.splitter = MoveSplitter(
+            config, self.gcode, fade_start, fade_end)
+        self.adj_stop = fade_end
+        if fade_end <= fade_start:
+            # Never Fade Mesh
+            self.adj_stop = self.FADE_DISABLE
+        self.gcode.register_command(
+            'BED_MESH_OUTPUT', self.cmd_BED_MESH_OUTPUT,
+            desc=self.cmd_BED_MESH_OUTPUT_help)
+        # Register G81 as alias
+        self.gcode.register_command(
+            'G81', self.cmd_BED_MESH_OUTPUT,
+            desc=self.cmd_BED_MESH_OUTPUT_help)
+        self.gcode.register_command(
+            'BED_MESH_CLEAR', self.cmd_BED_MESH_CLEAR,
+            desc=self.cmd_BED_MESH_CLEAR_help)
+        self.gcode.set_move_transform(self)
+    def printer_state(self, state):
+        if state == 'connect':
+            self.toolhead = self.printer.lookup_object('toolhead')
+    def set_mesh(self, mesh):
+        # Assign the current mesh.  If set to None, no transform
+        # is applied
+        self.z_mesh = mesh
+        self.splitter.set_mesh(mesh)
+        if mesh is None:
+            self.z_adjust = 0.
+    def get_position(self):
+        # Return last, non-transformed position
+        if self.z_mesh is None:
+            # No mesh calibrated, so send toolhead position
+            return self.toolhead.get_position()
+        else:
+            # return current position minus the current z-adjustment
+            x, y, z, e = self.toolhead.get_position()
+            return [x, y, z - self.z_adjust, e]
+    def move(self, newpos, speed):
+        gc_z = newpos[2] - self.gcode.base_position[2]
+        if self.z_mesh is None or gc_z >= self.adj_stop:
+            # No mesh calibrated, or mesh leveling phased out.
+            self.z_adjust = 0.
+            self.toolhead.move(newpos, speed)
+        else:
+            prev_pos = self.get_position()
```

Comment:
FYI, the previous position will always be the last position sent to move() or the last position returned by get_position().  The toolhead and gcode classes cache last position as an optimization - so it is safe to cache it in the move transform class as well.

Reply:
Okay, I can do that.